services/gen_job/gen_job.py

```diff
@@ -0,0 +1,30 @@
+from util import DictObj, createLogger
+from .prompt import generate_job_prompt
+from inference import inference
+
+logger = createLogger("job_expression_generator")
+
+class Payload(DictObj):
+    api_key: str = ""
+    adaptor: str = ""
+    instruction: str = ""
+    state: dict = {}
+    expression: str = ""
+
+def main(dataDict) -> str:
+    data = Payload(dataDict)
+    logger.info("Running job expression generator with gpt3_turbo model")
+
+    result = generate("gpt3_turbo", data.api_key, data.adaptor, data.instruction, data.state, data.expression)
```

Comment:
Sure, for now the output for:
```
{
"existing_expression": "",
   "adaptor": "@openfn/language-dhis2@4.0.3",
   "state":" { 'data': { 'name': 'bukayo saka', 'gender': 'male' } }",
   "instruction": "Create a new trackedEntityInstance 'person' in dhis2 for the 'dWOAzMcK2Wt' orgUnit."
}
```

is returned as:

![Screenshot 2024-08-08 203216](https://github.com/user-attachments/assets/d0d5d5bd-7df9-428d-84ed-93dadc8f37ab)

This is quite close to what is expected but there is still scope for improvement. What do you think about this?

Reply:
Yes, there is a lot of scope for improvement here. We can improve the code quality by improving the RAG service but how do we plan to add the attribute IDs to the prompt? 
Is this something we can expect:
![Screenshot 2024-08-11 214640](https://github.com/user-attachments/assets/2fa87db1-19ab-4e60-86de-62aefbc82b9e)
 
Also, what if we plan to deploy a fine-tuned model to OpenAI and call that model? This would increase the quality of response significantly.

Reply:
Attribute IDs are an interesting problem to solve.

We can't add DHIS2 specific rules to the prompt. Maybe we can add some prompt rules to use placeholders if it doesn't know real values.

We do have some helper APIs to get lists of attribute IDs for a given dhis2 server... but that's a lot of work to set up, and it still gives us a dhis2 specific prompt.

I think placeholders, like in your screenshot, are the best we can hope for at the moment

Reply:
Fine tuning our own model is something we want to explore but is outside the scope of this project.